### PR TITLE
add rspec test, argument validation for product and  warehouse

### PR DESCRIPTION
**Pull request by bpietraga**

### CODE REVIEW

File: shop_app/lib/product.rb [hunk 1]
```diff
@@ -3,19 +3,17 @@ class Product
   @@next_id = 0
 
   def initialize(name, price, vat = 23)
+    check_price_argument(price)
```
Comment (pawurb) [inline shop_app/lib/product.rb]: you could call it check_price_argument! to show that it raises an error. this is the convention

File: shop_app/lib/product.rb [hunk 2]
```diff
@@ -3,19 +3,17 @@ class Product
...
   private
-  def check_product_arguments
-    unless @name != nil || is_a?(String)
-      raise ArgumentError
-    end
-    unless @price != nil || @price > 0
-      raise ArgumentError
-    end
+  def check_price_argument(price)
+    raise ArgumentError unless price.is_a?(Numeric)
```
Comment (pawurb) [inline shop_app/lib/product.rb]: you could:
```
raise ArgumentError, "Error message" unless price.is_a?(Numeric)
```
to have more descriptive error when debugging etc

File: shop_app/Guardfile [hunk 3]
```diff
@@ -1,9 +1,8 @@
 # A sample Guardfile
 # More info at https://github.com/guard/guard#readme
 
-## Uncomment and set this to only include directories you want to watch
-# directories %w(app lib config test spec features) \
-#  .select{|d| Dir.exists?(d) ? d : UI.warning("Directory #{d} does not exist")}
+directories %w(spec) \
+select{|d| Dir.exists?(d) ? d : UI.warning("Directory #{d} does not exist")}
```
Comment (pawurb) [inline shop_app/Guardfile]: maybe user do end syntax and multiline. this is not very readable:
```
directories(%w(spec)).do |directory|
...
end
```

File: shop_app/spec/unit/product_spec.rb [hunk 4]
```diff
@@ -1,11 +1,34 @@
...
 
-  it "check butter price" do
-    expect(Product.new("Butter", 12, 10).price).to eql(12)
+  context "#price" do
+    it "return 12 when given given" do
+      expect(Product.new("Butter", 12, 10).price).to eql(12)
+    end
+
+    it "price must be > 0" do
+      expect{Product.new("Test" , -1)}.to raise_error ArgumentError
```
Comment (pawurb) [inline shop_app/spec/unit/product_spec.rb]: always add space before and after { }

File: shop_app/spec/unit/warehouse_spec.rb [hunk 5]
```diff
@@ -2,6 +2,34 @@
...
+
+  it "test product creation" do
+    @warehouse = Warehouse.new("Warehouse", [])
+
+    expect(@warehouse).to be_a Warehouse
+  end
+
+  context ".add" do
+    it "product creation" do
+      @warehouse = Warehouse.new("Warehouse", [])
```
Comment (pawurb) [inline shop_app/spec/unit/warehouse_spec.rb]: i think test description does not match

File: shop_app/spec/unit/warehouse_spec.rb [hunk 6]
```diff
@@ -2,6 +2,34 @@
...
+
+      expect(@warehouse).to be_a Warehouse
+    end
+  end
+
+  context "#name" do
+    it "name cannot be nil" do
+      expect{Warehouse.new(nil , [])}.to raise_error ArgumentError
```
Comment (smt116) [inline shop_app/spec/unit/warehouse_spec.rb]: missing spacebars around curly brackets

File: shop_app/spec/unit/warehouse_spec.rb [hunk 7]
```diff
@@ -2,6 +2,34 @@
...
+    @warehouse = Warehouse.new("Warehouse", [])
+
+    expect(@warehouse).to be_a Warehouse
+  end
+
+  context ".add" do
+    it "product creation" do
+      @warehouse = Warehouse.new("Warehouse", [])
+
+      expect(@warehouse).to be_a Warehouse
```
Comment (smt116) [inline shop_app/spec/unit/warehouse_spec.rb]: sometimes you are using `method foo` and another time `method(foo)` for assertions. Be consistent. I would use version with brackets.

File: shop_app/spec/unit/product_spec.rb [hunk 8]
```diff
@@ -1,11 +1,34 @@
 require_relative '../../lib/app.rb'
+require_relative '../../lib/product.rb'
 
 RSpec.describe Product do
-  it "generate Butter product" do
-    expect(Product.new("Butter", nil, nil).name).to eql("Butter")
+
+  it "test product creation" do
+    @product = Product.new("TestProduct", 1)
```
Comment (smt116) [inline shop_app/spec/unit/product_spec.rb]: instance variable is not needed here

Comment (smt116) [inline shop_app/spec/unit/product_spec.rb]: Basically normal variable has scope limited to the current context (local scope) while instance variable has "class scope". It is too wide for your use case (you need it only in a single spec, right?)

File: shop_app/Gemfile [hunk 9]
```diff
@@ -5,4 +5,5 @@ ruby "2.3.0"
 group :test do
   gem 'rspec'
   gem 'guard'
+  gem 'guard-rspec'
```
Comment (smt116) [inline shop_app/Gemfile]: guard should be a part of `development` group instead of `test`.

File: shop_app/lib/app.rb [hunk 10]
```diff
@@ -5,29 +5,35 @@
...
-            ]
+class App
+  def self.call
+    products = [Product.new("Butter",       2.14, 23),
+                Product.new("Shake",        3.0,  7),
+                Product.new("Papier",       9.99, 20),
+                Product.new("Mayo",         2.59, 23),
+                Product.new("Driller",      3.56, 7),
+                Product.new("Laser",        9.01, 20)
+                ]
```
Comment (smt116) [inline shop_app/lib/app.rb]: 👍 for aligning values inside `Product::new`, but (1) you have too many spacebars before second argument and (2) I would align also `)` at the end.

I think that there is something wrong with indentation. Use of of those versions:

```ruby
[
  Product.new(...),
  Product.new,
  ...
]
```

or

```ruby
[ Product.new,
  Product.new,
  ... ]
```

File: shop_app/lib/app.rb [hunk 11]
```diff
@@ -5,29 +5,35 @@
...
 
-generated_warehouse = Warehouse.new("Generated Warehouse", products)
+    generated_warehouse = Warehouse.new("Generated Warehouse", products)
```
Comment (smt116) [inline shop_app/lib/app.rb]: Do you really need `generated` prefix?

File: shop_app/lib/product.rb [hunk 12]
```diff
@@ -3,19 +3,21 @@ class Product
...
+  def check_price_argument!(price)
+    unless price.is_a?(Numeric)
+      raise ArgumentError, "Invalid price argument is not a number."
     end
-    unless @price != nil || @price > 0
-      raise ArgumentError
+
+    unless price >= 0
+      raise ArgumentError, "Invalid argument, price is < 0."
     end
```
Comment (smt116) [inline shop_app/lib/product.rb]: those checks could be also done in a single line: `raise ArgumentError, "..." unless something`

File: shop_app/spec/unit/product_spec.rb [hunk 13]
```diff
@@ -1,11 +1,33 @@
 require_relative '../../lib/app.rb'
+require_relative '../../lib/product.rb'
 
 RSpec.describe Product do
-  it "generate Butter product" do
-    expect(Product.new("Butter", nil, nil).name).to eql("Butter")
+  it "test product creation" do
+    product = Product.new("TestProduct", 1)
+    expect(product).to be_a Product
```
Comment (smt116) [inline shop_app/spec/unit/product_spec.rb]: This test and its description is confusing for me. Do you really need to check if `Product.new` returns `Product` object?

File: shop_app/spec/unit/product_spec.rb [hunk 14]
```diff
@@ -1,11 +1,33 @@
...
+      expect(Product.new("Butter", 1).name).to eql("Butter")
+    end
+  end
+
+  context "#price" do
+    it "return 12 when given given" do
+      expect(Product.new("Butter", 12, 10).price).to eql(12)
+    end
+
+    it "price must be > 0" do
```
Comment (smt116) [inline shop_app/spec/unit/product_spec.rb]: wuut? maybe `requires price greater than 0` instead?

File: shop_app/spec/unit/product_spec.rb [hunk 15]
```diff
@@ -1,11 +1,33 @@
...
+      expect { Product.new("Test" , -1) }.to raise_error ArgumentError
+    end
+
+    it "price cannot be nil" do
+      expect { Product.new("Test" , nil) }.to raise_error ArgumentError
+    end
+
+    it "price must by Numeric" do
+      expect { Product.new("Test" , "test") }.to raise_error ArgumentError
+    end
```
Comment (smt116) [inline shop_app/spec/unit/product_spec.rb]: this spec is confusing. Its description says that product's price has to be numeric (why with capital letter?), but in fact it checks if price cannot be a string.

I would change the whole group to something like this:

```ruby
describe "#price" do
  ...

  context "when price is set to nil" do
    it "raises exception" do
      expect { Product.new ... }.to raise_error(ArgumentError)
    end
  end

  context "when price is a string" do
  ...
```

You have also some messages when exceptions are raised so you can test them too.

Another thing is that `descibe "#price"` **is not a good place for testing those exceptions**. They won't be raised on `Product#price` call, but on the `Product#new` right?

File: shop_app/spec/unit/warehouse_spec.rb [hunk 16]
```diff
@@ -2,6 +2,26 @@
 require_relative '../../lib/warehouse.rb'
 
 RSpec.describe Warehouse do
+  let(:empty_warhouse) { Warehouse.new(nil , nil) }
```
Comment (smt116) [inline shop_app/spec/unit/warehouse_spec.rb]: It is used only in one place (line 14) so there is no point in extracting it here.

File: shop_app/spec/unit/warehouse_spec.rb [hunk 17]
```diff
@@ -2,6 +2,26 @@
...
+
+  it "test if new Warehouse object equals Warehouse class object" do
+    new_warehouse = Warehouse.new("Warehouse", [])
+    expect(new_warehouse).to be_a Warehouse
+  end
```
Comment (smt116) [inline shop_app/spec/unit/warehouse_spec.rb]: imho this spec is useless. Similar to `expect("foo").to be_a(String)` or `expect(1).to eq(1)` :P

File: shop_app/spec/unit/warehouse_spec.rb [hunk 18]
```diff
@@ -2,6 +2,26 @@
...
+
+  context "#name" do
+    it "name cannot be nil" do
+      expect { empty_warhouse }.to raise_error ArgumentError
+    end
```
Comment (smt116) [inline shop_app/spec/unit/warehouse_spec.rb]: The same thing like in https://github.com/bpietraga/ragnarson_intern/pull/1/files#r70180893

File: shop_app/lib/app.rb [hunk 19]
```diff
@@ -7,12 +7,13 @@
...
 class App
   def self.call
-    products = [Product.new("Butter",       2.14, 23),
-                Product.new("Shake",        3.0,  7),
-                Product.new("Papier",       9.99, 20),
-                Product.new("Mayo",         2.59, 23),
-                Product.new("Driller",      3.56, 7),
-                Product.new("Laser",        9.01, 20)
+    products = [
+                Product.new("Butter",   2.14,   23  ),
```
Comment (smt116) [inline shop_app/lib/app.rb]: wrong indentation.

File: shop_app/lib/product.rb [hunk 20]
```diff
@@ -3,15 +3,19 @@ class Product
...
     @name = name
     @price = price
     @vat = vat
   end
 
   private
-  def check_price_argument!(price)
+  def check_product_argument!(name, price)
+    unless name.is_a?(String)
+      raise ArgumentError, "Invalid name, argument is not a number."
```
Comment (smt116) [inline shop_app/lib/product.rb]: number?

File: shop_app/spec/unit/product_spec.rb [hunk 21]
```diff
@@ -2,32 +2,69 @@
...
+        expect(Product.new("Butter", 1).name).to eql("Butter")
+      end
+    end
+
+    context "when name is set to nil" do
+      it "raises exeption" do
+        expect {
+          empty_product = Product.new(nil , nil)
+          Product.name
+        }.to raise_error ArgumentError,
```
Comment (smt116) [inline shop_app/spec/unit/product_spec.rb]: exception is raised on `Product#name` call or on `Product#new`?

File: shop_app/spec/unit/product_spec.rb [hunk 22]
```diff
@@ -2,32 +2,69 @@
...
 
-  context "#name" do
-    it "returns proper value" do
-      expect(Product.new("Butter", 1).name).to eql("Butter")
+    context "when name is a integer" do
+      it "raises exeption" do
+        expect {
+          empty_product = Product.new(nil , nil)
+          Product.name
+        }.to raise_error ArgumentError,
```
Comment (smt116) [inline shop_app/spec/unit/product_spec.rb]: same here

Comment (smt116) [inline shop_app/spec/unit/product_spec.rb]: and in all assertions below

File: shop_app/lib/app.rb [hunk 23]
```diff
@@ -14,7 +14,7 @@ def self.call
                 Product.new("Mayo",     2.59,   23  ),
                 Product.new("Driller",  3.56,   7   ),
                 Product.new("Laser",    9.01,   20  )
-                ]
+               ]
```
Comment (smt116) [inline shop_app/lib/app.rb]: Two spaces for elements indentation:

```ruby
foo = [
         "foo",
         "bar"
      ]

Comment (smt116) [inline shop_app/lib/app.rb]: or

```ruby
foo = [ "foo",
        "bar" ]

File: shop_app/lib/app.rb [hunk 24]
```diff
@@ -7,14 +7,12 @@
 
 class App
   def self.call
-    products = [
-                Product.new("Butter",   2.14,   23  ),
+    products =[ Product.new("Butter",   2.14,   23  ),
```
Comment (smt116) [inline shop_app/lib/app.rb]: spacebar before `[`

**Pull request merged**